cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/WrappedSQLEngineCollection.java

```diff
@@ -161,12 +162,39 @@ public boolean tryStoreDirect(StageSpec stageSpec) {
     return wrapped.tryStoreDirect(stageSpec);
   }
 
+  @Override
+  public Set<String> tryMultiStoreDirect(PhaseSpec phaseSpec, Set<String> sinks) {
+    return wrapped.tryMultiStoreDirect(phaseSpec, sinks);
+  }
+
   @Override
   public Runnable createMultiStoreTask(PhaseSpec phaseSpec,
                                        Set<String> group,
                                        Set<String> sinks,
                                        Map<String, StageStatisticsCollector> collectors) {
-    return unwrap().createMultiStoreTask(phaseSpec, group, sinks, collectors);
+    return new Runnable() {
+      @Override
+      public void run() {
+        // Copy all sinks and groups into a ConcurrentHashSet set
+        Set<String> remainingGroups = new HashSet<>(group);
+        Set<String> remainingSinks = new HashSet<>(sinks);
+
+        // Invoke multi store direct task on wrapped collection
+        Set<String> consumedSinks = wrapped.tryMultiStoreDirect(phaseSpec, sinks);
+
+        // Removed all consumed sinks by the Multi store task from this group.
+        remainingGroups.removeAll(consumedSinks);
```

Comment:
minor nit: maybe add comment here saying some of the group stages can also be sink, hence we need to remove sinks from groups as well?

Reply:
Yes, the `group` contains all stages in the phase (including transforms).